internal/core/src/common/ChunkWriter.h

```diff
@@ -0,0 +1,184 @@
+#pragma once
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <vector>
+#include "arrow/array/array_primitive.h"
+#include "common/ChunkTarget.h"
+#include "arrow/array/array_base.h"
+#include "arrow/record_batch.h"
+#include "common/Chunk.h"
+#include "common/FieldDataInterface.h"
+namespace milvus {
+
+class ChunkWriterBase {
+ public:
+    ChunkWriterBase() = default;
+
+    ChunkWriterBase(File& file, size_t offset)
+        : file_(&file), file_offset_(offset) {
+    }
+
+    virtual void
+    write(std::shared_ptr<arrow::RecordBatchReader> data) = 0;
+
+    virtual std::shared_ptr<Chunk>
+    finish() = 0;
+
+    std::pair<char*, size_t>
+    get_data() {
+        return target_->get();
+    }
+
+ protected:
+    int row_nums_ = 0;
+    File* file_ = nullptr;
+    size_t file_offset_ = 0;
+    std::shared_ptr<ChunkTarget> target_;
+};
+
+template <typename ArrowType, typename T>
+class ChunkWriter : public ChunkWriterBase {
+ public:
+    ChunkWriter(int dim) : dim_(dim) {
+    }
+
+    ChunkWriter(int dim, File& file, size_t offset)
+        : ChunkWriterBase(file, offset), dim_(dim){};
+
+    void
+    write(std::shared_ptr<arrow::RecordBatchReader> data) override {
+        // chunk layout: data1, data2, ..., datan
+        auto size = 0;
+        auto row_nums = 0;
+        auto batch_vec = data->ToRecordBatches().ValueOrDie();
+
+        for (auto batch : batch_vec) {
+            row_nums += batch->num_rows();
+            auto data = batch->column(0);
+            auto array = std::dynamic_pointer_cast<ArrowType>(data);
+            size += array->length() * dim_ * sizeof(T);
```

Comment:
We may need additional handling on types of length less than 1, e.g., VECTOR_BINARY.

Reply:
> We may need additional handling on types of length less than 1, e.g., VECTOR_BINARY.

fixed